platform/unix/love.1
.\" (c) 2008-2011 Miriam Ruiz <little_miry@yahoo.es>
.\" (c) 2013 Bart van Strien <bart.bes@gmail.com>
.\"
.\" This software is provided 'as-is', without any express or implied warranty. In no event will the authors be held liable for any damagesarising from the use of this software.
.\"
.\" Permission is granted to anyone to use this software for any purpose, including commercial applications, and to alter it and redistribute it freely, subject to the following restrictions:
.\"
.\" 1. The origin of this software must not be misrepresented; you must not claim that you wrote the original software. If you use this software in a product, an acknowledgment in the product documentation would be appreciated but is not required.
.\"
.\" 2. Altered source versions must be plainly marked as such, and must not be misrepresented as being the original software.
.\"
.\" 3. This notice may not be removed or altered from any source distribution.
.\"
.\" Modifications:
.\" - Update version to 0.9 and remove reference to doc dir
.\" - Add fused and version flags

.TH "LÖVE" "1" "0.9" "" ""
.SH "NAME"
love \- 2D game development framework

.SH "SYNOPSIS"
.B love
[--fused] <\fIgame.love\fR>
.PP
.B love
--version
.PP

.SH "DESCRIPTION"
LÖVE was created to be a user\-friendly engine in which simple (or complicated) games could be made without having extensive knowledge of system or graphics functions and without having to dedicate time towards developing the same engine features time and time again.
.P
Developed with cross\-platform implementation in mind, it utilizes the latest open source libraries to deliver a similar game experience, independent of operating system. By relying on the Lua scripting language for game\-specific programming, it allows even the novice game creator to quickly and efficiently develop an idea into a fully working game.

.SH "SEE ALSO"
You can find more information at \fIhttp://love2d.org/\fR


changes.txt
LOVE 0.9.1 [Baby Inspector]
---------------------------

  Released: N/A

  * Added Source:clone.
  * Added ParticleSystem:clone.
  * Added love.graphics.setWireframe for debugging.
  * Added Mesh:setDrawRange and Mesh:getDrawRange.
  * Added instancing support to Meshes with Mesh:setInstanceCount.
  * Added CircleShape:getPoint and CircleShape:setPoint.
  * Added Mesh/SpriteBatch/ParticleSystem:setTexture, accepts Canvases and Images.
  * Added high-dpi window support for Retina displays in OS X, via the 'highdpi' window flag.
  * Added love.window.getPixelScale.
  * Added love.graphics.getSystemLimit.
  * Added ParticleSystem:moveTo, has smoother emitter movement compared to setPosition.
  * Added antialiasing support to Canvases.
  * Added Canvas:getFSAA.
  * Added 'love_ScreenSize' built-in variable in shaders.
  * Added love.getVersion.
  * Added support for gamma-correct rendering.
  * Added love.graphics.isSupported("srgb").
  * Added love.math.gammaToLinear and love.math.linearToGamma.

  * Deprecated Mesh/SpriteBatch/ParticleSystem:setImage.
  * Deprecated love.graphics.getMaxImageSize and love.graphics.getMaxPointSize.

  * Fixed love.graphics.scale with negative values causing incorrect line widths.
  * Fixed Joystick:isDown using 0-based button index arguments.
  * Fixed Source:setPitch to error when infinity or NaN is given.
  * Fixed love.graphics.setCanvas() to restore the proper viewport and scissor rectangles.
  * Fixed TrueType font glyphs which request a monochrome bitmap pixel mode.
  * Fixed love.graphics.reset causing crashes when called in between love.graphics.push/pop.
  * Fixed tab characters ("\t") to display properly with love.graphics.print.

  * Renamed love.graphics.getMaxImageSize to love.graphics.getMaxTextureSize (old function still exists.)

  * Updated the error text for love.filesystem’s module searchers when require fails.
  * Updated the love.filesystem module searchers to be tried after package.preload instead of before.
  * Updated love.graphics.newParticleSystem, newSpriteBatch, and newMesh to accept Canvases.
  * Updated Canvas drawing code, texture coordinates are no longer flipped vertically.
  * Updated Canvas:renderTo to work properly if a Canvas is currently active.
  * Updated ParticleSystem:setEmissionRate to accept non-integer numbers.
  * Updated Source:play to return a boolean indicating success.
  * Updated t.console in conf.lua to create the console before modules are loaded in Windows.

LOVE 0.9.0 [Baby Inspector]
---------------------------

	Released: 2013-12-13

  * Added better multiplayer networking support via ENet.
  * Added --fused command line argument, to simulate fusing.
  * Added liblove.
  * Added the ability to have exit values.
  * Added exit value of 1 in case of error by default.
  * Added basic support for the file:// uri scheme.
  * Added love.filesystem.isFused.
  * Added love.filesystem.getIdentity.
  * Added love.filesystem.append.
  * Added love.filesystem.getSize.
  * Added love.filesystem.mount and love.filesystem.unmount.
  * Added optional file search order parameter to love.filesystem.setIdentity.
  * Added File:isOpen and File:getMode.
  * Added Fie:setBuffer, File:getBuffer, and File:flush.
  * Added textinput event for unicode text input.
  * Added love.keyboard.setTextInput and love.keyboard.hasTextInput.
  * Added previously internal Rasterizer and GlyphData object methods.
  * Added support for UTF-8 ImageFonts.
  * Added Font:getAscent/getDescent/getBaseline.
  * Added Font:setFilter/getFilter.
  * Added Font:hasGlyphs.
  * Added angle, scale, and shear parameters to love.graphics.printf.
  * Added HDR canvas support.
  * Added mipmapping support (has isSupported test).
  * Added vertex shader support.
  * Added boolean support to Shader:send.
  * Added Canvas:getPixel.
  * Added blend mode "replace".
  * Added line join modes.
  * Added Mesh objects, allowing for arbitrary textured polygons.
  * Added multiple render target support to love.graphics.setCanvas.
  * Added love.graphics.setColorMask.
  * Added love.graphics.origin.
  * Added love.graphics.getRendererInfo.
  * Added love.graphics.getMaxImageSize.
  * Added SpriteBatch:getCount and SpriteBatch:getBufferSize.
  * Added SpriteBatch:getColor.
  * Added ParticleSystem:emit.
  * Added ParticleSystem:setInsertMode and ParticleSystem:getInsertMode.
  * Added many ParticleSystem getter methods.
  * Added DXT compressed texture support via love.image.newCompressedData.
  * Added love.image.isCompressed and Image:isCompressed.
  * Added Image/Canvas/ImageData:getDimensions.
  * Added anisotropic filtering support for Images, Canvases, and Fonts.
  * Added Image:refresh.
  * Added Image:getData.
  * Added SoundData:getDuration and SoundData:getSampleCount.
  * Added Source:isPlaying.
  * Added Source:setRelative and Source:isRelative.
  * Added Source:setCone and Source:getCone.
  * Added Source:getChannels.
  * Added new Channels API for love.thread.
  * Added limited table support to Channel:push.
  * Added Thread:getError.
  * Added Thread:isRunning.
  * Added threaderror event.
  * Added love.math module.
  * Added a platform-independent (good) random implementation to love.math.
  * Added RandomGenerator objects.
  * Added BezierCurve objects.
  * Added love.math.triangulate and love.math.isConvex.
  * Added love.math.noise.
  * Added love.timer.getAverageDelta.
  * Added Data:getString.
  * Added Contact:getChildren.
  * Added love.system module.
  * Added love.system.getClipboardText and love.system.setClipboardText.
  * Added love.system.getOS and love.system.getProcessorCount.
  * Added love.window module.
  * Added love.window.isVisible.
  * Added flags to love.window.setMode.
  * Added monitor choosing support to love.window.setMode.
  * Added support for resizable, borderless, and non-centered windows.
  * Added support for "fullscreen-desktop" mode.
  * Added window resize and visible events.
  * Added love.window.getDimensions.

src/modules/love/love.cpp
/**
 * Copyright (c) 2006-2014 LOVE Development Team
 *
 * This software is provided 'as-is', without any express or implied
 * warranty.  In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 **/

// LOVE
#include "common/config.h"
#include "common/version.h"
#include "common/runtime.h"

#include "love.h"

#ifdef LOVE_WINDOWS
#include <windows.h>
#endif // LOVE_WINDOWS

#ifdef LOVE_LEGENDARY_CONSOLE_IO_HACK
#include <fcntl.h>
#include <io.h>
#include <iostream>
#include <fstream>
#endif // LOVE_LEGENDARY_CONSOLE_IO_HACK

// Libraries.
#ifdef LOVE_ENABLE_LUASOCKET
#	include "libraries/luasocket/luasocket.h"
#endif
#ifdef LOVE_ENABLE_ENET
#	include "libraries/enet/lua-enet.h"
#endif

// Scripts
#include "scripts/boot.lua.h"

// All modules define a c-accessible luaopen
// so let's make use of those, instead
// of addressing implementations directly.
extern "C"
{
#if defined(LOVE_ENABLE_AUDIO)
	extern int luaopen_love_audio(lua_State*);
#endif
#if defined(LOVE_ENABLE_EVENT)
	extern int luaopen_love_event(lua_State*);
#endif
#if defined(LOVE_ENABLE_FILESYSTEM)
	extern int luaopen_love_filesystem(lua_State*);
#endif
#if defined(LOVE_ENABLE_FONT)
	extern int luaopen_love_font(lua_State*);
#endif
#if defined(LOVE_ENABLE_GRAPHICS)
	extern int luaopen_love_graphics(lua_State*);
#endif
#if defined(LOVE_ENABLE_IMAGE)
	extern int luaopen_love_image(lua_State*);
#endif
#if defined(LOVE_ENABLE_JOYSTICK)
	extern int luaopen_love_joystick(lua_State*);
#endif
#if defined(LOVE_ENABLE_KEYBOARD)
	extern int luaopen_love_keyboard(lua_State*);
#endif
#if defined(LOVE_ENABLE_MATH)
	extern int luaopen_love_math(lua_State*);
#endif
#if defined(LOVE_ENABLE_MOUSE)
	extern int luaopen_love_mouse(lua_State*);
#endif
#if defined(LOVE_ENABLE_PHYSICS)
	extern int luaopen_love_physics(lua_State*);
#endif
#if defined(LOVE_ENABLE_SOUND)
	extern int luaopen_love_sound(lua_State*);
#endif
#if defined(LOVE_ENABLE_SYSTEM)
	extern int luaopen_love_system(lua_State*);
#endif
#if defined(LOVE_ENABLE_TIMER)
	extern int luaopen_love_timer(lua_State*);
#endif
#if defined(LOVE_ENABLE_THREAD)
	extern int luaopen_love_thread(lua_State*);
#endif
#if defined(LOVE_ENABLE_WINDOW)
	extern int luaopen_love_window(lua_State*);
#endif
	extern int luaopen_love_boot(lua_State*);
}

static const luaL_Reg modules[] = {
#if defined(LOVE_ENABLE_AUDIO)
	{ "love.audio", luaopen_love_audio },
#endif
#if defined(LOVE_ENABLE_EVENT)
	{ "love.event", luaopen_love_event },
#endif
#if defined(LOVE_ENABLE_FILESYSTEM)
	{ "love.filesystem", luaopen_love_filesystem },
#endif
#if defined(LOVE_ENABLE_FONT)
	{ "love.font", luaopen_love_font },
#endif
#if defined(LOVE_ENABLE_GRAPHICS)
	{ "love.graphics", luaopen_love_graphics },
#endif
#if defined(LOVE_ENABLE_IMAGE)
	{ "love.image", luaopen_love_image },
#endif
#if defined(LOVE_ENABLE_JOYSTICK)
	{ "love.joystick", luaopen_love_joystick },
#endif
#if defined(LOVE_ENABLE_KEYBOARD)
	{ "love.keyboard", luaopen_love_keyboard },
#endif
#if defined(LOVE_ENABLE_MATH)
	{ "love.math", luaopen_love_math },
#endif
#if defined(LOVE_ENABLE_MOUSE)
	{ "love.mouse", luaopen_love_mouse },
#endif
#if defined(LOVE_ENABLE_PHYSICS)
	{ "love.physics", luaopen_love_physics },
#endif
#if defined(LOVE_ENABLE_SOUND)
	{ "love.sound", luaopen_love_sound },
#endif
#if defined(LOVE_ENABLE_SYSTEM)
	{ "love.system", luaopen_love_system },
#endif
#if defined(LOVE_ENABLE_TIMER)
	{ "love.timer", luaopen_love_timer },
#endif
#if defined(LOVE_ENABLE_THREAD)
	{ "love.thread", luaopen_love_thread },
#endif
#if defined(LOVE_ENABLE_WINDOW)
	{ "love.window", luaopen_love_window },
#endif
	{ "love.boot", luaopen_love_boot },
	{ 0, 0 }
};

#ifdef LOVE_LEGENDARY_CONSOLE_IO_HACK
int w__openConsole(lua_State *L);
#endif // LOVE_LEGENDARY_CONSOLE_IO_HACK

const char *love_version()
{
	return love::VERSION;
}

const char *love_codename()
{
	return love::VERSION_CODENAME;
}

static int w_love_getVersion(lua_State *L)
{
	lua_pushinteger(L, love::VERSION_MAJOR);
	lua_pushinteger(L, love::VERSION_MINOR);
	lua_pushinteger(L, love::VERSION_REV);
	lua_pushstring(L, love::VERSION_CODENAME);
	return 4;
}

int luaopen_love(lua_State * L)
{
	love::luax_insistglobal(L, "love");

	// Set version information.
	lua_pushstring(L, love::VERSION);
	lua_setfield(L, -2, "_version");

	lua_pushnumber(L, love::VERSION_MAJOR);
	lua_setfield(L, -2, "_version_major");
	lua_pushnumber(L, love::VERSION_MINOR);
	lua_setfield(L, -2, "_version_minor");
	lua_pushnumber(L, love::VERSION_REV);
	lua_setfield(L, -2, "_version_revision");

	lua_pushstring(L, love::VERSION_CODENAME);
	lua_setfield(L, -2, "_version_codename");

#ifdef LOVE_LEGENDARY_CONSOLE_IO_HACK
	lua_pushcfunction(L, w__openConsole);
	lua_setfield(L, -2, "_openConsole");
#endif // LOVE_LEGENDARY_CONSOLE_IO_HACK

	lua_newtable(L);

	for (int i = 0; love::VERSION_COMPATIBILITY[i] != 0; ++i)
	{
		lua_pushstring(L, love::VERSION_COMPATIBILITY[i]);
		lua_rawseti(L, -2, i+1);
	}

	lua_setfield(L, -2, "_version_compat");

	lua_pushcfunction(L, w_love_getVersion);
	lua_setfield(L, -2, "getVersion");

#ifdef LOVE_WINDOWS
	lua_pushstring(L, "Windows");
#elif defined(LOVE_MACOSX)
	lua_pushstring(L, "OS X");
#elif defined(LOVE_LINUX)
	lua_pushstring(L, "Linux");
#else
	lua_pushstring(L, "Unknown");
#endif
	lua_setfield(L, -2, "_os");

	// Preload module loaders.
	for (int i = 0; modules[i].name != 0; i++)
	{
		love::luax_preload(L, modules[i].func, modules[i].name);
	}

#ifdef LOVE_ENABLE_LUASOCKET
	love::luasocket::__open(L);
#endif
#ifdef LOVE_ENABLE_ENET
	love::luax_preload(L, luaopen_enet, "enet");
#endif

	return 1;
}

#ifdef LOVE_LEGENDARY_CONSOLE_IO_HACK

int w__openConsole(lua_State *L)
{
	static bool is_open = false;
	if (is_open)
		return 0;
	is_open = true;

	if (GetConsoleWindow() != NULL || AllocConsole() == 0)
		return 0;

	const int MAX_CONSOLE_LINES = 5000;
	CONSOLE_SCREEN_BUFFER_INFO console_info;

	// Set size.
	GetConsoleScreenBufferInfo(GetStdHandle(STD_OUTPUT_HANDLE), &console_info);
	console_info.dwSize.Y = MAX_CONSOLE_LINES;
	SetConsoleScreenBufferSize(GetStdHandle(STD_OUTPUT_HANDLE), console_info.dwSize);

	SetConsoleTitle(TEXT("LOVE Console"));

	FILE * fp;

	// Redirect stdout.
	fp = freopen("CONOUT$", "w", stdout);
	if (L && fp == NULL)
		luaL_error(L, "Console redirection of stdout failed.");

	// Redirect stdin.
	fp = freopen("CONIN$", "r", stdin);
	if (L && fp == NULL)
		luaL_error(L, "Console redirection of stdin failed.");

	// Redirect stderr.
	fp = freopen("CONOUT$", "w", stderr);
	if (L && fp == NULL)
		luaL_error(L, "Console redirection of stderr failed.");

	return 0;
}

#endif // LOVE_LEGENDARY_CONSOLE_IO_HACK

int luaopen_love_boot(lua_State *L)
{
	if (luaL_loadbuffer(L, (const char *)love::boot_lua, sizeof(love::boot_lua), "boot.lua") == 0)
		lua_call(L, 0, 1);

	return 1;
}



